### ISSUE TITLE
calculate_3b_upper packs the wrong index triangle

### ISSUE THREAD

**Issue by Irratzo** [StefanoSanvitoGroup/MLdensity#11]
Connected:

- Branch [fix-2b-upper-packing](https://github.com/StefanoSanvitoGroup/MLdensity/tree/fix-2b-upper-packing)
- Pull request #12
- (Independent of #6 / PR #7 — see the last section)

## Summary

`calculate_3b_upper` in `jlgridfingerprints/src/jlcontraction.pyx:159-165` iterates the
symmetric radial index pair over the **lower** triangle:

```cython
for i1 in range(nmax):
    for i2 in range(nmax):
        if (i1 >= i2):                                          # lower triangle
            i = <int>(i1*nmax - (i1*(i1-1)/2) + (i2-i1) + i3*k1) # upper-triangle packing
```

The packing formula on the second line is the standard bijection of the **upper** triangle
onto a flat array, valid only for $i_2 \ge i_1$. Fed the lower triangle, the term $(i_2 - i_1)$
is non-positive and walks the write pointer *backwards* into the block belonging to a smaller
$i_1$.

The result is neither a permutation nor a deduplication. Some slots are assigned twice — and
`jac[i] = s` is an assignment, not an accumulation, so the later, larger-$i_1$ pair **silently
replaces** the value already written there, and that earlier pair then appears nowhere in the
output. An equal number of slots are never assigned at all and keep the zero prefill from
`jlcontraction.pyx:154-155`.

Concretely, at $n_{\max} = 5$ radial orders (the `aluminium` example), one angular block of 15
slots comes out like this:

```
slot        0    1    2    3    4    5    6    7    8    9   10   11   12   13   14
now       0,0    .    .    .  0,1  1,1    .  0,2  1,2  0,3  0,4  1,4  2,4  3,4  4,4
correct   0,0  0,1  0,2  0,3  0,4  1,1  1,2  1,3  1,4  2,2  2,3  2,4  3,3  3,4  4,4
           ok   --   --   --    x   ok   --    x    x    x    x    x    x   ok   ok
```

`n,m` is the radial pair whose value the slot ends up holding and `.` is a slot left at exactly
zero. The last row summarises: `ok` means the slot holds what it should, `--` that it is dead,
and `x` that it holds some *other* pair's value. Four of fifteen slots are dead, and **four of the fifteen radial pairs — (1,3),
(2,2), (2,3) and (3,3) — never reach the output at all.** The eleven that survive are all
present, but at indices that do not match the convention `get_3b_index` documents.

One thing this does *not* mean: no *numerically distinct* value is computed and then destroyed
by a different one. Because `calculate_3b_upper` is called as
`calculate_3b_upper(jacobi_ni_2b, jacobi_ni_2b, legendre_lij)` with the same expansion in both
slots and a symmetric Legendre matrix, the contraction is exactly symmetric,
$s(i_1,i_2) = s(i_2,i_1)$. So the lower triangle computes the same fifteen numbers the upper one
would. The damage is entirely in the addressing: four of those numbers are written to a slot
that a later iteration then overwrites, so they are lost, and four slots are left at zero.

With $m$ the number of Jacobi orders actually returned, per angular order $\ell$ there are
$U(m) = m(m+1)/2$ slots of which

$$D(m) = \left\lfloor \frac{(m-1)^2}{4} \right\rfloor$$

are permanently zero, and exactly $D(m)$ distinct radial pairs are lost to overwrite.
$D/U \to 1/2$ as $m$ grows, so the loss worsens along the radial axis.

**Diagonal pairs are among those lost** — $(2,2)$ at $m = 4$, for instance. No symmetry
argument can justify discarding a pair with itself, which is what rules out reading the
current behaviour as deliberate redundancy under the block's $n_1 \leftrightarrow n_2$
symmetry.

Across a whole descriptor the pattern repeats once per angular order. For `aluminium`
($n_{\max} = [15, 6]$, `lmax` 6, `double_shifted`, so $m = 5$ and 120 features):

```
block       columns      identically zero
----------- ------------ ----------------
1B          0 - 14       none
2B  l=0     15 - 29      16 17 18 21
2B  l=1     30 - 44      31 32 33 36
2B  l=2     45 - 59      46 47 48 51
2B  l=3     60 - 74      61 62 63 66
2B  l=4     75 - 89      76 77 78 81
2B  l=5     90 - 104     91 92 93 96
2B  l=6     105 - 119    106 107 108 111
```

Those indices are measured, not predicted: they are the columns found identically zero across
every grid point of every `aluminium` structure. The same four offsets — 1, 2, 3 and 6 within the
block — repeat in all seven angular blocks, 28 columns in total, 23.3 % of the descriptor. The 1B
block and any heterospecies 2B block are untouched.

The fix is one character: `if (i2 >= i1)`. That restores the convention the function's own
docstring, the packing formula, and `get_3b_index` all already state.

### ELI5

The loop hands the packer its radial pairs in mirror-image order from the one its filing
formula expects. So the packer files some pairs into pigeonholes that already belong to other
pairs — the earlier occupant is thrown away, unrecorded — and leaves the pigeonholes at the end
of each row empty forever. The descriptor still comes out exactly the advertised length. Part
of it is just permanently blank, and part of what should have gone there was dropped on the
floor.

## How to read this issue

- **1B / 2B** — the library's grid-centred convention. The **2B (two-body)** term is the
  radial-angular expansion over *pairs* of neighbours; it is computed by the functions named
  `*_3b` for historical, atom-centred reasons.
- **same-species / heterospecies block** — a 2B block pairs two neighbour species. When the
  two are the same species the pair matrix is symmetric and only its triangle is stored
  (`calculate_3b_upper`); when they differ, the full matrix is stored (`calculate_3b`).
  **Only the same-species path is affected.**
- **$m$** — the number of Jacobi radial orders `expand_jacobi` actually returns:
  $m = n_{\max,2b} - 1$ under `double_shifted`, else $m = n_{\max,2b}$.
- **dead column** — a descriptor column that is identically zero at every grid point of every
  structure, by construction rather than by chance.
- **budgeted vs effective width** — budgeted is what `number_of_features()` returns; effective
  is how many columns can ever be non-zero.

## 1. A check you can run in one line, on a model you already have

This needs no rebuild, no retraining, and no acceptance of anything else in this issue. Load the
`aluminium` model this repository ships and look at its fitted coefficients:

```python
import pickle, numpy as np
m = pickle.load(open("aluminium/ml_model/train_ml/scikit_linear_model_chg.p", "rb"))
print(np.abs(m.coef_).max(), np.linalg.norm(m.coef_))
```

Write $\mathbf{c}$ for that coefficient vector — one entry per descriptor column, so 120 entries
here. The two printed numbers are

- $\max_j |c_j|$, the **largest single coefficient in absolute value** — `np.abs(m.coef_).max()`;
- $\lVert\mathbf{c}\rVert_2 = \big(\sum_j c_j^2\big)^{1/2}$, the **Euclidean norm of the whole
  coefficient vector** — `np.linalg.norm(m.coef_)`.

They are the right things to look at because a linear model's coefficients live on the scale of
"target per unit of feature". A single coefficient thousands of times larger than the rest is not
a physical result; it is the signature of a solver that was asked to invert something singular.

The shipped model gives $\max_j |c_j| = 1.97\times 10^{4}$ and
$\lVert\mathbf{c}\rVert_2 = 3.35\times 10^{4}$. Refit the same example after the fix and the
entire vector has norm $2.06\times 10^{-2}$ — **six orders of magnitude smaller.**

### Why zero columns produce enormous coefficients rather than zero ones

This is the part that reads backwards at first, so it is worth spelling out.

The dead **columns of the design matrix** are zero. The **coefficients attached to them** are
not, and there is no reason they should be. For a column that is identically zero, *every*
coefficient value yields exactly the same prediction — the fit is completely indifferent to it,
and the training data contains no information whatsoever about what it should be. Something
still has to come out of the solver, and what comes out is whatever its arithmetic produces from
floating-point rounding noise in a direction where the true singular value is zero. Here that is
$\sim 10^{4}$.

Splitting the shipped vector by column shows this exactly: all but a part in $10^{12}$ of
$\sum_j c_j^2$ sits on the 28 dead columns, and all 20 coefficients above $10^3$ are
dead-column coefficients. The 92 live coefficients sum in quadrature to $2.08\times10^{-2}$ —
essentially identical to the *entire* vector of the refitted model ($2.06\times 10^{-2}$). So
the six orders of magnitude are not a difference in the fitted physics at all. They are 28
arbitrary numbers riding along in the parameter vector.

Predictions are unaffected, because a coefficient of $2\times 10^{4}$ multiplying a column that
is exactly zero contributes exactly zero. That is the whole reason this has stayed invisible: it
shows up in the parameters, never in the output.

*Why the solver does not simply discard those directions:* `Ridge(solver='svd')` drops a
direction only when its singular value falls below an **absolute** threshold of $10^{-15}$. The
dead directions never reach it — rounding leaves them near $\epsilon\,\sigma_{\max}$, which is
$\approx 10^{-7}$ for `aluminium` and $\approx 10^{-4}$ for `benzene`, because those design
matrices have norms of $10^{9}$ and $10^{12}$. So they are inverted rather than truncated, and
`assert np.isfinite(model.coef_).all()` passes because the numbers are finite, not because they
are sane.

## 2. Scale, on this repository's own examples

All four examples set `double_shifted: True`, so $m = n_{\max,2b} - 1$.

| example | species | `nmax` | `lmax` | m | budgeted width | effective width | dead | dead, of each same-species 2B block |
|---|---|---|---|---|---|---|---|---|
| `aluminium` | 1 | [15, 6] | 6 | 5 | 120 | 92 | 28 | 26.7 % |
| `molybdenium` | 1 | [20, 12] | 11 | 11 | 812 | 512 | 300 | 37.9 % |
| `benzene` | 2 | [27, 12] | 5 | 11 | 1572 | 1272 | 300 | 37.9 % |
| `2d_mos2` | 2 | [18, 11] | 10 | 10 | 2346 | 1906 | 440 | 36.4 % |

*Legend.* **species** — number of chemical species. **`nmax`** — `[nmax_1b, nmax_2b]` from each
example's `create_data.py`. **budgeted width** — what `number_of_features()` returns and what
the paper quotes; unchanged by the fix. **effective width** — columns that can ever be
non-zero. **dead** — the difference. The last column gives the dead share of one same-species
2B block, which is larger than the share of the whole descriptor because the 1B and
heterospecies blocks are unaffected.

The budgeted column reproduces all four published counts exactly, which is what makes the
effective column checkable. We confirmed the dead counts by generating real descriptors on the
real structures in each example: 28 / 300 / 300 / 440, matching the closed form exactly.

## 3. What it costs, measured

Two legs differing only in the guard character, retrained from the shipped CHGCARs in one
SLURM job on one node, identical thread pinning, identical sampled points. Full detail is in
PR #12.

| | aluminium | benzene |
|---|---|---|
| dead columns, before → after | 28 → 0 | 300 → 0 |
| numerical rank of the training design matrix, before → after | 92 → 120 | 1088 → 1269 |
| training RMSE | −7.06 % | −2.83 % |
| test RMSE | **−6.97 %** | +0.54 % |
| test MAE | −6.68 % | −0.39 % |
| test MaxAE | −10.09 % | −1.04 % |

*Legend.* Negative is an improvement in the error rows. **numerical rank** —
`numpy.linalg.matrix_rank` of the training design matrix: how many linearly independent
directions the fit actually has to work with. **The rank rising is the expected direction, not a
typo.** A column that is identically zero contributes no direction, so before the fix a
120-column `aluminium` matrix could supply at most 92 — and supplied exactly 92, meaning the
surviving columns were all independent and the entire deficit was the dead ones. Afterwards all
120 columns are live and independent. Benzene gains fewer directions than columns
(300 recovered, rank +181) because some of the restored pairs are close to combinations of ones
already present.

Aluminium improves on every metric in both partitions. Benzene improves on training
everywhere and on two of three test metrics; its test RMSE rises 0.54 %, which we traced to
the unregularized `alpha=0.0` fit rather than to the basis — at any non-zero ridge penalty the
restored basis is ahead on benzene too. We report the +0.54 % rather than tuning it away.

A pre-registered check that could have refuted us: the broken kernel's occupied slots hold a
strict subset of the true radial pairs, so the baseline's column span is nested inside the
fixed one, and for an unregularized least-squares fit training error therefore **cannot** rise.
Had it risen, our reading would have been wrong. It fell on both examples.

## 4. Why we think this went unnoticed

Not to make a point about the code, but because "surely someone would have seen this" is the
first reasonable objection, and all three of these are checkable:

1. **The advertised length is correct.** `number_of_features()` counts the full upper triangle
   and the kernel writes into a buffer of exactly that length. Every descriptor length ever
   printed or published is right; nothing downstream sees a shape mismatch.
2. **The fit absorbs it in silence**, in the way described in §1 — no exception, no warning, no
   `NaN`. A normal-equations or Cholesky solver would have raised a singular-matrix error on
   day one; the SVD path does not.
3. **Nothing tested the index arithmetic.** The suite covers polynomials, shapes and round
   trips. No test asserted the packing was a bijection, and none inspected a descriptor vector
   for structural zeros. PR #12 adds both.

## 5. What we are asking for

A review of the one-character guard change and its tests in PR #12. We are not asking for
anything else: `number_of_features()` is unchanged, no public API moves, and the heterospecies
kernel is untouched.

One consequence needs stating plainly, because it is silent: **descriptor values and column
order both change, while the feature count does not.** No shape error will fire. A stored
model from an earlier version applies its coefficients to a different basis, so its
predictions are meaningless — descriptors must be regenerated and models refitted. Published
accuracy figures were correctly measured against the basis that was actually in use; they are
not wrong, they simply describe a narrower basis than their stated feature count implies.

## Independence from #6 / PR #7, and merge order

Different function, different statement, different mechanism; neither fix touches the other's
code path. The branch is based on `fix-alpha-beta-truncation` rather than `stable` so that
merging this does not drop the alpha/beta fix, and PR #12 therefore targets that branch —
GitHub will retarget it to `stable` automatically when PR #7 merges. Under the version rule
agreed in issue #8, the branch claims 0.1.9 as the next unused patch version above PR #7's
0.1.8; both numbers move if an unrelated PR merges ahead.

## Why an outside caller is reporting this

The downstream project [vimp-prediction](https://iffgit.fz-juelich.de/phd-project-wasmer/projects/vimp-prediction)
uses `jlgridfingerprints` to learn a KKR-formalism electron *potential* rather than a charge
density. A rank analysis of its own design matrix showed a fixed fraction of the 2B block
identically zero at every grid point, which is what led to the kernel. We are reporting it
because it is a defect in the library as it stands, independent of what we are doing with it.
